inveniosoftware/react-invenio-deposit
# funding: add new funding/award field

closes inveniosoftware/invenio-app-rdm#763

#### TODO:
- [x] Funding modals:
  - [x] Standard award
    - [x] No results link to custom award form
    - [x] Pagination ⚠️ (Can't figure out how to set the RSK query state `page` and `size` properly)
  - [x] Custom award
- [x] Polish UI/commented out code 
- [x] Remove old components


#### Missing stuff
- Test out and adapt when the backend is ready. At the moment all the data it's being saved on the formik's metadata field.
- "Add award" modal:
  - Search is already sending the appropriate data to the backend (`&q=foo`) but it couldn't be tested properly
  - Filter by funder is just a UI select field at the moment, whose values are being fetched from the mock API (funders that have at least one award linked). This endpoint has to be crafted, a good idea raised by Lars is to serve it under `/api/vocabularies/funders?awards=1` or `/api/vocabularies/funders?hasawards=true`. To be decided and done. Afterwards, a way to filter awards by funder must be implemented. Idea: `/api/vocabularies/awards?funder_id=00k4n6c32`
  - Pagination RSK component is included but is never showing up. This is due to its `currentSize` and `currentPage` props are `-1`, which is the default value. These props should be feed by the `state.query` but for some reason is not working as expected. Might be related to the mock api.
- UI polishing:
  - Prettify radio buttons in "Add award" modal.
  - Prettify "Filter by funder" select field in "Add award" modal.

#### To test

- Check out all PRs linked to inveniosoftware/invenio-app-rdm#763 and run as `--watch-module`.

⚠️ The vocabularies APIs for `/awards` and `/funders` are yet not implemented.

In order to test this PR let's mock the API response. There are multiple ways to do this, I've used the [`EasyMock`](https://marketplace.visualstudio.com/items?itemName=EasyMock.easymock) VS Code extension that allows you to create dummy API pretty quick. My configuration file is:

```js
// example.js
module.exports = {
  '/api/vocabularies/awards': {
    hits: {
      hits: [
        {
          id: 'a88e6ee0-106c-4413-b3e8-11024bef24ff',
          title: 'OpenAIRE - Open Access Infrastructure for Research in Europe',
          number: '246686',
          link: 'https://cordis.europa.eu/project/id/246686',
          funder: {
            id: '00k4n6c32',
            name: 'European Commission (EU)',
          },
        },
        {
          id: 'awardA',
          title: 'CANCER &AIDS DRUGS--PRECLIN PHARMACOL/TOXICOLOGY',
          number: 'N01CM037835-016',
          link: 'https://cancer.aids.org/toxicology',
          funder: {
            id: 'xyz',
            name: 'National Institutes of Health (US)',
          },
        },
        {
          title:
            'Beyond the Standard Model at the LHC and with Atom Interferometers.',
          number: '228169',
          id: 'awardB',
          funder: {
            id: 'xyz',
            name: 'National Institutes of Health (US)',
          },
        },

        {
          title: 'Environmental Conditions in Glaucoma Patients',
          number: '747441',
          id: 'awardC',
          funder: {
            id: 'xyz',
            name: 'National Institutes of Health (US)',
          },
        },
      ],
      total: 4,
    },
    links: {
      next:
        'https://inveniordm.web.cern.ch/api/vocabularies/awards?page=2&size=25&sort=bestmatch',
      self:
        'https://inveniordm.web.cern.ch/api/vocabularies/awards?page=1&size=25&sort=bestmatch',
      prev:
        'https://inveniordm.web.cern.ch/api/vocabularies/awards?page=1&size=25&sort=bestmatch',
    },
    sortBy: 'bestmatch',
  },

  // All funders (potentially thousands)
  '/api/vocabularies/funders': {
    hits: {
      hits: [
        {
          id: '00k4n6c32',
          name: 'European Commission',
        },
        {
          id: 'xyz',
          name: 'National Institutes of Health (US)',
        },
        {
          id: 'aaa',
          name: 'Yet Another Funder',
        },
        {
          id: 'bbb',
          name: 'Super Funder',
        },
      ],
      total: 4,
    },
    links: {
      next:
        'https://inveniordm.web.cern.ch/api/vocabularies/funders?page=2&size=25&sort=bestmatch',
      self:
        'https://inveniordm.web.cern.ch/api/vocabularies/funders?page=1&size=25&sort=bestmatch',
      prev:
        'https://inveniordm.web.cern.ch/api/vocabularies/funders?page=2&size=25&sort=bestmatch',
    },
    sortBy: 'bestmatch',
  },

  // Funders that have awards linked
  '/api/vocabularies/funders/awards': {
    hits: {
      hits: [
        {
          id: '00k4n6c32',
          name: 'European Commission',
        },
        {
          id: 'xyz',
          name: 'National Institutes of Health (US)',
        },
      ],
      total: 2,
    },
    sortBy: 'bestmatch',
  },
};

```

By default, API is served under `http://127.0.0.1:9999`. As the default CSP config does not permit to fetch content from this URL, we need to manually allow it. Add to `my-site/invenio.cfg` this line:

```python
# For mock API
APP_DEFAULT_SECURE_HEADERS['content_security_policy']['default-src'].append("http://127.0.0.1:9999")
```
